linux/ibus-keyman/src/engine.c

```diff
@@ -576,40 +627,95 @@ ibus_keyman_engine_process_key_event (IBusEngine     *engine,
         switch(action_items[i].type)
         {
             case KM_KBP_IT_CHAR:
-                numbytes = g_unichar_to_utf8(action_items[i].character, utf8);
-                if (numbytes > 12)
-                {
-                    g_error("g_unichar_to_utf8 overflowing buffer");
+                keyman->marker = False;
+                if (g_unichar_type(action_items[i].character) == G_UNICODE_SURROGATE) {
+                    if (keyman->firstsurrogate == 0) {
+                        keyman->firstsurrogate = action_items[i].character;
+                        g_message("first surrogate %d", keyman->firstsurrogate);
+                    }
+                    else {
+                        glong items_read, items_written;
+                        gunichar2 utf16_pair[2] = {keyman->firstsurrogate, action_items[i].character};
+                        gchar *utf8_pair = g_utf16_to_utf8 (utf16_pair, 2,
+                            &items_read,
+                            &items_written,
+                            NULL);
+                        if (keyman->char_buffer == NULL) {
+                            keyman->char_buffer = utf8_pair;
+                        }
+                        else {
+                            gchar *new_buffer = g_strjoin("", keyman->char_buffer, utf8_pair, NULL);
+                            g_free(keyman->char_buffer);
+                            g_free(utf8_pair);
+                            keyman->char_buffer = new_buffer;
+                        }
+                        keyman->firstsurrogate = 0;
+                    }
                 }
-                if (numbytes)
-                {
-                    utf8[numbytes] = 0;
+                else {
+                    gchar *utf8 = (gchar *) g_new0(gchar, 12);
+                    numbytes = g_unichar_to_utf8(action_items[i].character, utf8);
+                    if (numbytes > 12) {
+                        g_error("g_unichar_to_utf8 overflowing buffer");
+                        g_free(utf8);
+                    }
+                    else {
+                        g_message("unichar:U+%04x, bytes:%d, string:%s", action_items[i].character, numbytes, utf8);
+                        if (keyman->char_buffer == NULL) {
+                            keyman->char_buffer = utf8;
+                        }
+                        else {
+                            gchar *new_buffer = g_strjoin("", keyman->char_buffer, utf8, NULL);
+                            g_free(keyman->char_buffer);
+                            g_free(utf8);
+                            keyman->char_buffer = new_buffer;
+                        }
+                    }
                 }
-                g_message("CHAR action unichar:U+%04x, bytes:%d, string:%s", action_items[i].character, numbytes, utf8);
-                ibus_keyman_engine_commit_string(keyman, utf8);
                 break;
             case KM_KBP_IT_MARKER:
+                keyman->marker = True;
```

Comment:
This behaviour for markers is not correct. A marker is a flag in the context (and output). There can be multiple markers in the context, and they have order and a specific position between any two characters. There may be multiple ordered markers between two characters,

Markers are flushed when context is reset. They do not persist. Reasons for context being reset include:

(a) focus changes - to another text store
(b) any mouse button event
(c) cursor movement - for text stores without a IBUS_CAP_SURROUNDING_TEXT capability
(d) when KM_KBP_IT_CONTEXT_INVALIDATE is received.

Reply:
changed keyman->marker from bool to gchar.
(a) is reset
(b) behaviour with mouse is trickier so I'm looking into it more in the next PR (linux-backspace branch)
(c) context resets with IBUS_CAP_SURROUNDING_TEXT as well
(d) reset

Reply:
The use of keyman->marker isn't to know anything about the markers but to keep track of what to do when you press the backspace key, or otherwise have a BACK action

So, looking at it more it looks like I should go back to the bool

Looking at test `020 - deadkeys and backspace.kmn`:
(1) and (4) a hanging deadkey or deadkeys will be deleted by BKSP
(2) char + deadkey - BKSP deletes both
(3) deadkey + char - BKSP deletes both
(5) char_a + dk + dk + char_b - BKSP deletes char_b and both dk
(6) char_a + dk + dk + dk + dk + char_b  deletes char_b and both dk

so for (1) and (4) the marker state should be TRUE after END
for the rest it should be FALSE so that the char is deleted

Are there any other deadkey scenarios that aren't covered that should be added to that test?

Reply:
I've renamed keyman->marker to hasonlymarker and made it a bool again. That can handle all the scenarios in test 020.